### PR TITLE
Refactor/throw exceptions refactor

### LINKED ISSUE #86: Implement correct Exceptions where needed.
- Decide when to use exception.
- Implement the correct exception where due.

### LINKED ISSUE #185: Change our Singleton classes to the "proper" Java Singletons
Also refactor Singleton classes to enums

### PR DESCRIPTION
**Pull request by OscarSanner**
will close #86 also closes #185

### CODE REVIEW

**Review by yenanw — APPROVED** [review]
looks gucchi

File: app/.idea/vcs.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="VcsDirectoryMappings">
+    <mapping directory="$PROJECT_DIR$/.." vcs="Git" />
+  </component>
+</project>
```
Comment (Puh00) [inline app/.idea/vcs.xml:6]: Remove?

File: app/.idea/misc.xml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="ExternalStorageConfigurationManager" enabled="true" />
+</project>
```
Comment (Puh00) [inline app/.idea/misc.xml:4]: Remove?

File: app/.idea/gradle.xml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        <option name="modules">
+          <set>
+            <option value="$PROJECT_DIR$" />
+          </set>
+        </option>
+        <option name="useQualifiedModuleNames" value="true" />
+      </GradleProjectSettings>
+    </option>
+  </component>
+</project>
```
Comment (Puh00) [inline app/.idea/gradle.xml:18]: Remove?

File: app/.idea/.gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
+/workspace.xml
```
Comment (Puh00) [inline app/.idea/.gitignore:2]: Remove?

**Review by Puh00 — APPROVED** [review]
lgtm but can we remove .idea/ before merging

**Review by Puh00 — APPROVED** [review]
Great!

**Pull request merged**